feat(accounts): implement transaction details screen

mena-rizkalla opened the pull request:
## Summary
Implemented the **Transaction Details** screen for Loan and Savings accounts, allowing users to view specific information about a transaction.

## Changes
- **Data Layer:**
  - Created `TransactionDetails` model to handle detailed fields (breakdown, receipt, balance).
  - Added single transaction fetch endpoints to Loan and Savings services.
  - Implemented logic to handle inconsistent API field names (`type` vs `transactionType`) using `@JsonNames`.
- **UI:**
  - Created `TransactionDetailsScreen` displaying:
    - Status (Success/Reversed) with color coding.
    - Payment Method, Receipt #, and Remarks.
    - Full breakdown for loans (Principal, Interest, Fees, Penalties).
    - Outstanding Balance.
- **Fixes:**
  - Fixed an issue where transaction list items in `TransactionScreen` were not clickable.
  - Updated navigation logic to pass `accountId` correctly.
- **Docs:**
  - Updated KDoc for navigation and ViewModel events.

## Fixes
Fixes - [MM-409](https://mifosforge.jira.com/browse/MM-409)

## Screenshots
| Before | After |
|--------|-------|
| Feature did not exist | <img src="https://github.com/user-attachments/assets/1ebbd63f-8043-4080-870e-c104b091e041" width="300" /> <br><br> <img src="https://github.com/user-attachments/assets/a82b9d48-a725-4b66-b5b7-3e2cd5c71061" width="300" /> |

## Checklist                             
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

[MM-409]: https://mifosforge.jira.com/browse/MM-409?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<table>
  <tr>
    <td width="33%">
      <video src="https://github.com/user-attachments/assets/6ce81315-1b93-4afb-8aae-10b9e45c4cdb" controls></video>
    </td>
    <td width="33%">
      <video src="https://github.com/user-attachments/assets/df96f9c8-b6c8-408e-ac3e-dd679366d689" controls></video>
    </td>
    <td width="33%">
      <video src="https://github.com/user-attachments/assets/069bb5c2-547a-4155-9cc3-801751376670" controls></video>
    </td>
  </tr>
</table>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Transaction details screen with full info (date, status, account ref, payment method, type, receipt, external ID) and retry/back handling
  * Clickable transaction items across lists that navigate to the details view
  * Transaction breakdown showing principal, interest, fees, penalties and remaining balance
  * Transaction details available for both savings accounts and loans

* **Documentation**
  * Added localized strings for transaction details and field labels

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mena-rizkalla committed:
feat(core): add transaction details fetching support

- Create TransactionDetails model to handle detailed fields.
- Add single transaction fetch endpoints to Loan and Savings services.
- Implement getTransactionDetails logic in Loan and Savings repositories.

feat(accounts): implement transaction details screen

- Create TransactionDetailsScreen to display status, payment breakdown, and receipt.
- Implement TransactionDetailsViewModel to fetch data based on account type.
- Integrate navigation from Transaction History list to TransactionDetails.
- Fix click event handling in TransactionScreen list items.

docs(accounts): update kdoc for navigation and viewmodel events

mena-rizkalla committed:
fix: address PR review comments (navigation safety, ui logic, model stability)

mena-rizkalla committed:
fix: transaction debit/credit logic in TransactionDetailsViewModel

- Updated logic for determining `isCreditValue` to check for substrings (withdrawal, disbursement, repayment, fee, charge, penalty, transfer) rather than exact matches.
- Added explicit checks for fee, charge, penalty, and outgoing transfers as debit transactions.

mena-rizkalla committed:
Formatting changes in TransactionDetailsScreen and TransactionDetailsViewModel

*   `TransactionDetailsScreen.kt`: Added a trailing comma for better formatting.
*   `TransactionDetailsViewModel.kt`: Removed an unnecessary empty line.

therajanmaurya commented on cmp-navigation/src/commonMain/kotlin/cmp/navigation/authenticated/AuthenticatedNavigation.kt:
```diff
@@ -163,6 +166,27 @@ internal fun NavGraphBuilder.authenticatedGraph(
             navigateBack = navController::popBackStack,
+            navigateToDetails = { transactionId ->
+
+                val route = runCatching {
```
why are you using runCatching? It blocks the thread, it is for testing not for production code.

therajanmaurya commented on cmp-navigation/src/commonMain/kotlin/cmp/navigation/authenticated/AuthenticatedNavigation.kt:
```diff
@@ -163,6 +166,27 @@ internal fun NavGraphBuilder.authenticatedGraph(
+
+                val route = runCatching {
+                    navController.currentBackStackEntry?.toRoute<AccountTransactionsNavRoute>()
+                }.getOrNull()
```
this should be done in AccountTransactoinNavigation

navController.currentBackStackEntry?.toRoute<AccountTransactionsNavRoute>()
                }.getOrNull()
                if (route != null &&
                    route.accountId > 0L &&
                    route.accountType in listOf(Constants.SAVINGS_ACCOUNT, Constants.LOAN_ACCOUNT)
                    
                    Passing the parmeter

therajanmaurya commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsScreen.kt:
```diff
@@ -0,0 +1,346 @@
+            val statusColor = if (transaction.status == "reversed") {
+                MaterialTheme.colorScheme.error
+            } else {
+                Color(0xFF2E7D32)
```
No hard coded color, check core/designsystem use kptTheme

therajanmaurya commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsScreen.kt:
```diff
@@ -0,0 +1,346 @@
+}
+
+@Composable
+fun TransactionDetailContent(
```
Create preview of the screen

therajanmaurya commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsViewModel.kt:
```diff
@@ -0,0 +1,190 @@
+    private fun fetchTransactionDetails() {
+        updateState { it.copy(uiState = ScreenUiState.Loading) }
+
+        loadJob?.cancel()
```
don't manage the job here, remove it and viewModelScope manages in viewmodel scope by it's self.

therajanmaurya commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsViewModel.kt:
```diff
@@ -0,0 +1,190 @@
+    private suspend fun loadSavingsTransaction() {
+        savingsRepository.getSavingsAccountTransactionDetails(state.accountId, state.transactionId)
+            .collect { dataState ->
+                handleDataState(dataState)
```
Error should ne handled here, you are handling in parent method that's wrong of handling error

therajanmaurya commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsViewModel.kt:
```diff
@@ -0,0 +1,190 @@
+    private fun TransactionDetails.toUiTransaction(): UiTransactionDetails {
+        val typeLower = this.type?.value?.lowercase().orEmpty()
+
+        val isCreditValue = when {
```
move this in TransactionDetails data class

therajanmaurya commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/TransactionDetails.kt:
```diff
@@ -0,0 +1,63 @@
+
+@Serializable
+@Parcelize
+data class TransactionDetails(
```
I can see we have transaction details response is

```
{
  "amount": 559.88,
  "currency": {
    "code": "USD",
    "decimalPlaces": 2,
    "displayLabel": "US Dollar ($)",
    "displaySymbol": "$",
    "name": "US Dollar",
    "nameCode": "currency.USD"
  },
  "date": "2025-12-13",
  "id": 3,
  "interestPortion": 559.88,
  "manuallyReversed": false,
  "type": {
    "code": "loanTransactionType.repayment",
    "contra": false,
    "description": "Repayment",
    "disbursement": false,
    "id": 2,
    "recoveryRepayment": false,
    "repayment": true,
    "repaymentAtDisbursement": false,
    "waiveCharges": false,
    "waiveInterest": false,
    "writeOff": false
  }
}
```

Can you cross check the API

therajanmaurya commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/TransactionDetails.kt:
```diff
@@ -0,0 +1,63 @@
+import org.mifos.mobile.core.model.entity.client.Type
+
+@Serializable
+@Parcelize
```
remove @Parcelize

mena-rizkalla committed:
Refactor Transaction Details UI, logic and navigation

- `TransactionDetailsScreen`: Update status colors to use `AppColors`, remove payment method and receipt fields, and add `DevicePreview`.
- `TransactionDetailsViewModel`: Move credit calculation logic to entity, simplify coroutine launching, and add error handling for data flows.
- `TransactionDetails`: Add `isCredit` computed property, remove `paymentDetailData` and `Parcelable` implementation.
- `AuthenticatedNavigation`: Update `accountTransactionsDestination` callback signature to explicitly pass account ID and type.
- `AccountsTransactionNavigation`: Move route argument parsing logic inside the composable destination.
- `SavingAccountsListService`: Correct the API endpoint path parameter for retrieving transaction details.

mena-rizkalla committed:
Implemented navigation from Recent Transactions to Transaction Details screen.
- Updated `RecentTransactionViewModel` and `RecentTransactionState` to handle `OnTransactionClick` action and emit `NavigateToDetails` events.
- Integrated navigation callback in `RecentTransactionScreen` and `RecentTransactionNavGraph` to route to the details view.
- Updated debit transaction icon tint to `AppColors.lightRed` in `TransactionDetailsScreen`.

biplab1 commented on feature/recent-transaction/src/commonMain/kotlin/org/mifos/mobile/feature/recent/transaction/viewmodel/RecentTransactionViewModel.kt:
```diff
@@ -39,24 +44,27 @@ class RecentTransactionViewModel(
 ) : ViewModel() {
 
-    private val _uiState = MutableStateFlow(
+    private val mutableUiState = MutableStateFlow(
```
Please follow the uiState naming convention as used in other existing feature modules. Update everywhere as required.

mena-rizkalla replied:
Other existing viewmodels use BaseViewModel unlike RecentTransactionViewmodel !!

And i almost naming it like baseViewmodel 

<img width="1080" height="257" alt="Screenshot_20251216-151506" src="https://github.com/user-attachments/assets/33ad2286-3c88-462d-ac07-1fbbdc7550f6" />

biplab1 commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/TransactionDetails.kt:
```diff
@@ -0,0 +1,72 @@
+    val officeName: String? = null,
+
+    @OptIn(ExperimentalSerializationApi::class)
+    @JsonNames("transactionType")
```
You could use this instead:

```suggestion
    @SerialName("transactionType")
```

Ref: [SerialName Official documentation](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-serial-name/)

biplab1 commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsScreen.kt:
```diff
@@ -0,0 +1,361 @@
+@DevicePreview
+@Composable
+@Suppress("UnusedPrivateMember")
+private fun TransactionDetailContentPreview() {
```
Private can be removed as done in other modules.

```suggestion
fun TransactionDetailContentPreview() {
```

biplab1 reviewed (changes requested):
Suggested some minor changes. Please upload screen recording showing the screen navigation.

biplab1 commented on feature/recent-transaction/src/commonMain/kotlin/org/mifos/mobile/feature/recent/transaction/viewmodel/RecentTransactionViewModel.kt:
```diff
@@ -120,13 +128,38 @@ class RecentTransactionViewModel(
+        }
+    }
+
+    private fun determineAccountType(account: Any): String {
```
`determineAccountType(account: Any)` could take a more specific type instead of `Any`

mena-rizkalla replied:
Okay ,but should i check or not as i can see the one who write RecentTransactionViewModel use only savingAccount repository

mena-rizkalla committed:
Merge branch 'development' into feature/accounts/transaction-details

mena-rizkalla committed:
Refactored RecentTransactionViewModel and updated TransactionDetails

- **RecentTransactionViewModel**: Renamed `mutableUiState` to `_uiState` to follow naming conventions. Simplified navigation logic by hardcoding `accountType` to `SAVINGS_ACCOUNT` and removing the unused `determineAccountType` helper.
- **TransactionDetails**: Replaced `@JsonNames` with `@SerialName` for the `transactionType` field.
- **TransactionDetailsScreen**: Changed `TransactionDetailContentPreview` visibility to public and updated the suppression annotation to `UnusedMember`.

biplab1 commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/TransactionDetails.kt:
```diff
@@ -0,0 +1,72 @@
+
+    val officeName: String? = null,
+
+    @OptIn(ExperimentalSerializationApi::class)
```
Is `@OptIn` still required for using SerialName? I don't think so. Please check.

biplab1 commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/transactionDetail/TransactionDetailsScreen.kt:
```diff
@@ -0,0 +1,361 @@
+
+        Spacer(modifier = Modifier.height(DesignToken.spacing.large))
+
+        // 2. Details Section
```
Can you tell me what is the purpose of such comments?

mena-rizkalla replied:
That was a placeholder, I will remove it.

biplab1 reviewed (approved):
Some further minor changes. Rest looks good to me. This can be merged after the requested updates have been made.

mena-rizkalla committed:
Removed comments and annotations from TransactionDetails screen and entity.
